clippy_lints/src/new_without_default.rs

```diff
@@ -48,12 +49,75 @@ declare_clippy_lint! {
     "`pub fn new() -> Self` method without `Default` implementation"
 }
 
+declare_clippy_lint! {
+    /// ### What it does
+    /// If a type has an auto-derived `Default` trait and a `fn new() -> Self`,
+    /// this lint checks if the `new()` method performs custom logic rather
+    /// than simply calling the `default()` method.
+    ///
+    /// ### Why is this bad?
+    /// Users expect the `new()` method to be equivalent to `default()`,
+    /// so if the `Default` trait is auto-derived, the `new()` method should
+    /// not perform custom logic.  Otherwise, there is a risk of different
+    /// behavior between the two instantiation methods.
+    ///
+    /// ### Example
+    /// ```no_run
+    /// #[derive(Default)]
+    /// struct MyStruct(i32);
+    /// impl MyStruct {
+    ///   fn new() -> Self {
+    ///     Self(42)
+    ///   }
+    /// }
+    /// ```
+    ///
+    /// Users are unlikely to notice that `MyStruct::new()` and `MyStruct::default()` would produce
+    /// different results. The `new()` method should use auto-derived `default()` instead to be consistent:
+    ///
+    /// ```no_run
+    /// #[derive(Default)]
+    /// struct MyStruct(i32);
+    /// impl MyStruct {
+    ///   fn new() -> Self {
+    ///     Self::default()
+    ///   }
+    /// }
+    /// ```
+    ///
+    /// Alternatively, if the `new()` method requires a non-default initialization, implement a custom `Default`.
```

Comment:
This would unify the `new` and `default` implementation, losing either one or the other. I think the docs should suggest renaming `new` to clarify the difference.

Reply:
@llogiq thx for review!  Could you clarify how you think the docs should change?  The Rust-lang issue that actually caused me to write this lint was that `new()` would set the separator to `' '`, whereas the `default()` set it to `'\0'`.  So unifying them **should** loose one or the other, and consolidate the code, but keep the original function as `new()`

Reply:
P.S. are there any blockers to start the FCP?

Reply:
> @llogiq thx for review! Could you clarify how you think the docs should change? The `rustc` issue that actually caused me to write this lint was that `new()` would set the separator to `' '`, whereas the `default()` set it to `'\0'`. So unifying them **should** loose one or the other, and consolidate the code, but keep the original function as `new()`

I understand it as "Another less standard method name would be better than `new()` if it is different from `default()`, you could suggest renaming the `new` implementation".

Reply:
thx @samueltardieu , fixed